Sources/DocCDocumentation/IndexStoreDB+Extensions.swift

```diff
@@ -0,0 +1,98 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+package import IndexStoreDB
+import SKLogging
+import SemanticIndex
+@_spi(LinkCompletion) import SwiftDocC
+
+extension CheckedIndex {
+  /// Find a `SymbolOccurrence` that is considered the primary definition of the symbol with the given `DocCSymbolLink`.
+  ///
+  /// If the `DocCSymbolLink` has an ambiguous definition, the most important role of this function is to deterministically return
+  /// the same result every time.
+  package func primaryDefinitionOrDeclarationOccurrence(
+    ofDocCSymbolLink symbolLink: DocCSymbolLink
+  ) -> SymbolOccurrence? {
+    var components = symbolLink.components
+    guard components.count > 0 else {
+      return nil
+    }
+    // Do a lookup to find the top level symbol
+    let topLevelSymbol = components.removeLast()
+    var topLevelSymbolOccurrences: [SymbolOccurrence] = []
+    forEachCanonicalSymbolOccurrence(byName: topLevelSymbol.name) { symbolOccurrence in
+      topLevelSymbolOccurrences.append(symbolOccurrence)
+      return true  // continue
+    }
+    topLevelSymbolOccurrences = topLevelSymbolOccurrences.filter {
+      let symbolInformation = LinkCompletionTools.SymbolInformation(fromSymbolOccurrence: $0)
+      return symbolInformation.matches(topLevelSymbol.disambiguation)
+    }
+    // Search each potential symbol's parents to find an exact match
+    let symbolOccurences = topLevelSymbolOccurrences.filter { topLevelSymbolOccurrence in
+      var components = components
+      var symbolOccurrence = topLevelSymbolOccurrence
+      while let parentSymbolOccurrence = symbolOccurrence.parent(self), !components.isEmpty {
+        let nextComponent = components.removeLast()
```

Comment:
Using `popLast` we don’t need the `isEmpty` check anymore. Also let’s do `popLast` first because it’s faster than an index lookup.

```suggestion
      while let nextComponent = components.popLast(), let parentSymbolOccurrence = symbolOccurrence.parent(self) {   
```

Reply:
Done.